sdk/turing/ensembler.py

```diff
@@ -26,9 +26,16 @@ def ensemble(
         :param features: pandas.Series, containing a single row with input features
         :param predictions: pandas.Series, containing a single row with all models predictions
                 `predictions['model-a']` will contain prediction results from the model-a
-        :param treatment_config: dictionary, representing the configuration of a treatment,
-                that should be applied to a given record. If the experiment engine is not configured
-                for this Batch experiment, then `treatment_config` will be `None`
+        :param treatment_config: Optional[pandas.Series], representing the configuration of a
+                treatment, that should be applied to a given record/payload.
+
+                For batch experiments, the `treatment_config` can be `None`. If an attribute
+                representing a treatment plan of all the treatments and their respective
+                configuration is set using the `initialize` method, the `treatment_config`
+                for any treatment can be accessed during all calls made to `ensemble`.
+
+                For real-time experiments, the `treatment_config` is either a pandas.Series
+                object, or None, if the experiment engine is not configured.
```

Comment:
@romanwozniak I'm not sure if this what you referred to by updating the interface with the changes to `treatment_config`. Basically it's now (soon) going to be a `pandas.Series` object containing treatment config info, just as how `features` and `predictions` contain data within as a `pandas.Series` object. If you were referring to something like a payload unpacking step for `treatment_config` because we're expecting a payload with a standard structure, that part's going to be handled by the future real-time ensembler engine.

Reply:
> If an attribute representing a treatment plan of all the treatments and their respective configuration is set using the `initialize` method, the `treatment_config` for any treatment can be accessed during all calls made to `ensemble`.

I'm not sure if I follow this. Can we for now don't touch `treatment_config` for batch experiments at all? This requires a much deeper understanding of how batch experiments are run and how to connect ensembling jobs with the experiment engine.

> Basically it's now (soon) going to be a pandas.Series object containing treatment config info, just as how features and predictions contain data within as a pandas.Series object. If you were referring to something like a payload unpacking step for treatment_config because we're expecting a payload with a standard structure, that part's going to be handled by the future real-time ensembler engine.

I can't remember saying, that `treatment_config` should necessarily be a `pandas.Series`. I think we'll get a better understanding of how this all clicks together once we are moving closer to the integration of all pieces together (basically, after pyfunc ensembler server is ready and after you'll be able to package it as a docker image and deploying it as a custom ensembler in Turing)

Reply:
> I'm not sure if I follow this. Can we for now don't touch treatment_config for batch experiments at all? This requires a much deeper understanding of how batch experiments are run and how to connect ensembling jobs with the experiment engine.

Okay right now we aren't touching anything involving the batch experiments, so there shouldn't be any issue with the current changes on batch ensembling. 

As to what I was trying to describe in what you quoted, I was having in mind this batch ensembling example I found which shows an implementation of an ensembler with the `PyFunc` interface. It looks something like this:
```python
class MyEnsembler(PyFunc):
    def initialize(self, artifacts: dict):
        # store information about all treatments
        import yaml
        with open(artifacts[ENSEMBLER_CONFIG_PATH_ARTIFACT_KEY], 'r') as f:
            self.config = yaml.safe_load(f)

    def ensemble(self, features: pd.Series, predictions: pd.Series, treatment_config: Optional[dict]) -> Any:
        treatment = features['treatment']
        # access the treatment config stored in the self.config attribute, relevant to the treatment 
        treatment_config = self.config['treatments'][treatment]  
        
        if 'model_name' in treatment_config:
            # do something prescribed by the stored treatment_config
        else:
            # do something default
```
Essentially I was trying to say that for batch experiments, the current way `treatment_config`s are stored is through an attribute that gets initialised.

> I can't remember saying, that treatment_config should necessarily be a pandas.Series. I think we'll get a better understanding of how this all clicks together once we are moving closer to the integration of all pieces together (basically, after pyfunc ensembler server is ready and after you'll be able to package it as a docker image and deploying it as a custom ensembler in Turing)

I think I was trying to say that `treatment_config` is going to be a pandas.Series, if `ensemble` is called by the real-time ensembling engine. And you're right I think the details will be clearer when I bring up the new implementation of the real-time ensembling web service.

Reply:
> Essentially I was trying to say that for batch experiments, the current way treatment_configs are stored is through an attribute that gets initialised.

This is not how it's expected to work in the long run, hence we should not include this into the codebase.
In the long run, the gap between experiment engine and batch ensembler should be bridged and treatment configs should be retrieved from the exp engine before ensembling happens, probably as a separate transformation on the spark DataFrame. i.e. something like this:
```
treatmentsDF = spark.DataFrame(source).apply(fetch_treatment_udf)

result = spart.DataFrame(source) \
  .join(predictionsDF) \
  .join(treatmentsDF) \
  .apply(ensemble_udf)
```

Reply:
Okay then I'll remove the part of the docstrings that mention about the batch ensembling `treatment_config` workaround. Thanks a lot for the clarification!